client/scripts/views/modals/onboarding_modal/index.ts

```diff
@@ -0,0 +1,206 @@
+          // methods in controllers/login.ts. this means that all
+          // cases should be covered (either the account comes from
+          // the backend and the address is also loaded via
+          // AddressInfo, or the account is created on the frontend
+          // and the id is available here).
+          let addressInfo = app.user.addresses
+            .find((a) => a.address === account.address && a.chain === account.chain.id);
+
+          if (!addressInfo && account.addressId) {
+            // TODO: add keytype
+            addressInfo = new AddressInfo(account.addressId, account.address, account.chain.id, undefined);
+            app.user.addresses.push(addressInfo);
+          }
+
+          // link the address to the community
+          try {
+            if (vnode.attrs.joiningChain
+                && !app.user.getRoleInCommunity({ account, chain: vnode.attrs.joiningChain })) {
+              await app.user.createRole({ address: addressInfo, chain: vnode.attrs.joiningChain });
+            } else if (vnode.attrs.joiningCommunity
+                       && !app.user.getRoleInCommunity({ account, community: vnode.attrs.joiningCommunity })) {
+              await app.user.createRole({ address: addressInfo, community: vnode.attrs.joiningCommunity });
+            }
+          } catch (e) {
+            // this may fail if the role already exists, e.g. if the address is being migrated from another user
+          }
+
+          // set the address as active
+          await setActiveAccount(account);
+          if (app.user.activeAccounts.filter((a) => isSameAccount(a, account)).length === 0) {
+            app.user.setActiveAccounts(app.user.activeAccounts.concat([account]));
+          }
+          // TODO: set the address as default
+        } catch (e) {
+          console.trace(e);
+          // if the address' role wasn't initialized correctly,
+          // setActiveAccount will throw an exception but we should continue
+        }
+
+        // TODO: need mixpanel integration?
+        if (onNext) onNext(account);
+        m.redraw();
+      } else {
+        // log in as the new user
+        await initAppState(false);
+        // load addresses for the current chain/community
+        if (app.community) {
+          await updateActiveAddresses(undefined);
+        } else if (app.chain) {
+          const chain = app.user.selectedNode
+            ? app.user.selectedNode.chain
+            : app.config.nodes.getByChain(app.activeChainId())[0].chain;
+          await updateActiveAddresses(chain);
+        } else {
+          notifyError('Signed in, but no chain or community found');
+        }
+        if (onNext) onNext(account);
+        m.redraw();
+      }
+    };
+
+    return m('.OnboardingModal', [
+      m(OnboardingProgressBar, { step: vnode.state.step }),
+      step === 0 ? m(OnboardingConnect, {
+        address,
+        onUseWallet: () => {
+          vnode.state.step = 1;
```

Comment:
It would be nice to have an enum because these steps feel arbitrary.

Reply:
fixed